### PR TITLE
In macros, no longer require space between operation and following semicolon

### PR DESCRIPTION
**Pull request by dennisschagt**
Fixes https://github.com/newsboat/newsboat/issues/702.

Actual fix is in the latest commit (https://github.com/newsboat/newsboat/commit/7f7c5366ce9942e3cb46208eba5f396619eacb26) but the preceding commits make that fix possible.

### CODE REVIEW

File: test/configparser.cpp [hunk 1]
```diff
@@ -9,6 +11,74 @@
 
 using namespace newsboat;
 
+class DummyConfigHandler : public ConfigActionHandler {
```
[the remark this answers is not in the document]
Comment (dennisschagt) [reply test/configparser.cpp]: Done. I also put the class into a anonymous namespace which should make sure it never collides with classes in other compilation units: https://github.com/newsboat/newsboat/pull/1119/commits/d071ceba5ddfb1e496f254b40d858106c11e7a9d

File: test/configparser.cpp [hunk 2]
```diff
@@ -9,6 +11,74 @@
...
+
+	SECTION("no quoting") {
+		cfgParser.parse_line("command-name", location);
+		cfgParser.parse_line("command-name arg1", location);
+		cfgParser.parse_line("command-name arg1 arg2", location);
+
+		REQUIRE(handler.history.size() == 3);
+		REQUIRE(handler.history[0].second == std::vector<std::string>({}));
+		REQUIRE(handler.history[1].second == std::vector<std::string>({"arg1"}));
+		REQUIRE(handler.history[2].second == std::vector<std::string>({"arg1", "arg2"}));
```
[the remark this answers is not in the document]
Comment (dennisschagt) [reply test/configparser.cpp]: Done (https://github.com/newsboat/newsboat/pull/1119/commits/dce647594aebecee082fe2d0ac12af69e66a9d0c), although I did restructure the tests afterwards so when rebasing, I plan to squash all of the `parse_line()` tests into 1 commit.

File: test/configparser.cpp [hunk 3]
```diff
@@ -35,9 +35,9 @@ TEST_CASE("parse_line(): Handles both lines with and without quoting",
 	const std::string location = "dummy-location";
 
 	SECTION("no quoting") {
-		cfgParser.parse_line("command-name", location);
-		cfgParser.parse_line("command-name arg1", location);
-		cfgParser.parse_line("command-name arg1 arg2", location);
+		cfgParser.parse_line(" command-name ", location);
```
[the remark this answers is not in the document]
Comment (dennisschagt) [reply test/configparser.cpp]: Good point. Added a specific test case for this in https://github.com/newsboat/newsboat/pull/1119/commits/b738f3ade4fd48c060bb7a16cccc31fa7cd37ed5

File: src/configactionhandler.cpp [hunk 4]
```diff
@@ -12,4 +12,10 @@ void ConfigActionHandler::handle_action(const std::string& action,
 	handle_action(action, tokens);
 }
 
+// Derived classes should override either of the handle_action() overloads
```
[the remark this answers is not in the document]
Comment (dennisschagt) [reply src/configactionhandler.cpp]: Done: https://github.com/newsboat/newsboat/pull/1119/commits/c9d9d8c7e978e133deac86c06df4c7e1d0959714

File: src/keymap.cpp [hunk 5]
```diff
@@ -721,52 +721,31 @@ void KeyMap::handle_action(const std::string& action, const std::string& params)
...
-		if (tmpcmd.op != OP_NIL) {
-			cmds.push_back(tmpcmd);
+			auto arguments = utils::tokenize_quoted(operation);
+			MacroCmd cmd;
+			cmd.op = get_opcode(command_name.value());
+			if (cmd.op == OP_NIL) {
+				throw ConfigHandlerException(strprintf::fmt(_("`%s' is not a valid operation"),
+						command_name.value()));
+			}
+			for (const auto& argument : arguments) {
```
[the remark this answers is not in the document]
Comment (dennisschagt) [reply src/keymap.cpp]: Oops, not my brightest moment when I wrote that :)
https://github.com/newsboat/newsboat/pull/1119/commits/dc9fa19bec36352905872dc957df5fc0973139a5

File: src/configactionhandler.cpp [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+#include "utils.h"
+
+namespace newsboat {
+
+// Derived classes can override this function to handle the `params` string without the default tokenization
+void ConfigActionHandler::handle_action(const std::string& action,
+	const std::string& params)
+{
+	std::vector<std::string> tokens = utils::tokenize_quoted(params);
```
[the remark this answers is not in the document]
Comment (dennisschagt) [reply src/configactionhandler.cpp]: Processed all 3 of your `const` remarked and a few more: https://github.com/newsboat/newsboat/pull/1119/commits/c0c27b96527c343d60a34b4bf67260ff454a601a

**Comment by dennisschagt** [comment]
I can rebase+autosquash and force-push when the changes are accepted.

> It looks like it's possible to add some regression tests for this, though. Just copying the examples from #702 (comment) should be enough, but please check that the test doesn't pass on the second-to-last commit.

Done and verified that it fails on the second-to-last commit: https://github.com/newsboat/newsboat/pull/1119/commits/ca516194a6352d1b099fa41274c8a96467e6469e

> I think it also makes sense to update the docs along with the code. Docs currently put a space after an op, but not after set, and I can recall at least one IRC user who missed that little detail. Removing the spaces should make the docs look more uniform and less confusing.

I overlooked it, will create another commit+push soon.

**Comment by dennisschagt** [comment]
> I think it also makes sense to update the docs along with the code. Docs currently put a space after an op, but not after set, and I can recall at least one IRC user who missed that little detail. Removing the spaces should make the docs look more uniform and less confusing.

Done.

Thanks for the review!

File: test/configparser.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Minoru) [reply test/configparser.cpp]: Oh, I wish I thought of anonymous namespaces when I ran into that problem earlier! I'll use that from now on. :+1:

File: test/configparser.cpp [hunk 7]
```diff
@@ -24,58 +26,110 @@ class DummyConfigHandler : public ConfigActionHandler {
...
+		REQUIRE_THROWS(cfgParser.parse_line("foo", location));
+		REQUIRE_THROWS(cfgParser.parse_line("foo arg1 arg2", location));
+	}
 
-		REQUIRE(handler.history.size() == 3);
-		REQUIRE(handler.history[0].second == std::vector<std::string>({}));
-		REQUIRE(handler.history[1].second == std::vector<std::string>({"arg1"}));
-		REQUIRE(handler.history[2].second == std::vector<std::string>({"arg1", "arg2"}));
+	SECTION("different combinations of (un)quoted commands/arguments have the same result") {
+		std::vector<std::string> inputs = {
```
Comment (Minoru) [inline test/configparser.cpp:47]: Can this (and other vectors below) be const? :P

Comment (dennisschagt) [reply test/configparser.cpp]: Done while rebasing/squashing: https://github.com/newsboat/newsboat/compare/de0dd8f03cdf6b8c14f04e9bd33a57dec4744d13..de634778c3d36f6356f93cc764fbc96c7edff345

Comment (Minoru) [reply test/configparser.cpp]: Thanks!

**Pull request merged**